ai_diffusion/ui/widget.py

```diff
@@ -590,25 +593,75 @@ def move_cursor_to_end(self):
             self._single.setCursorPosition(len(self._single.text()))
 
 
+# SpinBox variant that allows manually entering strength values,
+# but snaps to model_steps on step actions (scrolling, arrows, arrow keys).
+class StrengthSpinBox(QSpinBox):
+    min_steps: int | None
+    max_steps: int | None
+
+    def __init__(self, parent=None):
+        super().__init__(parent)
+        # safe default
+        self.min_steps = None
+        self.max_steps = None
+        # for manual input
+        self.setMinimum(1)
+        self.setMaximum(100)
+
+    def step_info(self, value: int | None = None, pretty: bool = False) -> tuple[int, int] | None:
+        if self.min_steps is None or self.max_steps is None:
+            return None
+        # See workflow.py:_apply_strength
+        strength = (value or self.value()) / 100
+        steps = self.max_steps * strength
+        # round is more accurate, but int ensures that round steps end up on round values
+        steps = int(steps) if pretty else round(steps)
+        adj_max_steps = self.max_steps
+        if self.min_steps and steps < self.min_steps:
+            steps = self.min_steps
+            adj_max_steps = math.floor(self.min_steps / strength)
+        return (steps, adj_max_steps)
+
+    def stepBy(self, steps):
+        current_info = self.step_info(pretty=True)
+        adj_value = max(1, min(100, self.value() + steps))
+        if current_info is not None:
+            # keep going until we hit a unique step
+            while self.step_info(value=adj_value, pretty=True) == current_info:
+                adj_value += 1 if steps > 0 else -1
+            current_info = self.step_info(value=adj_value, pretty=True)
+            # keep going to the start of the section
+            if steps < 0:
+                while self.step_info(value=adj_value - 1, pretty=True) == current_info:
+                    adj_value -= 1
+        self.setValue(adj_value)
+
+
 class StrengthWidget(QWidget):
+    _model: Model | None
+    # whether we're in a live widget or quality widget
+    # determines which step size we use for model bounds
+    _is_live: bool | None
     value_changed = pyqtSignal(float)
 
-    def __init__(self, slider_range: tuple[int, int] = (1, 100), parent=None):
+    def __init__(
+        self, is_live: bool | None = None, slider_range: tuple[int, int] = (1, 100), parent=None
+    ):
         super().__init__(parent)
         self._layout = QHBoxLayout()
         self._layout.setContentsMargins(0, 0, 0, 0)
         self.setLayout(self._layout)
 
+        self._is_live = is_live
+        self._model = None
+
```

Comment:
You can also connect `settings.changed` event here if you want to have immediate update when switching step display.

Example handler: https://github.com/Acly/krita-ai-diffusion/blob/main/ai_diffusion/ui/region.py#L321

Reply:
oh hey nice. done